src/liballoc/alloc.rs

```diff
@@ -111,25 +119,19 @@ unsafe fn exchange_malloc(size: usize, align: usize) -> *mut u8 {
         align as *mut u8
     } else {
         let layout = Layout::from_size_align_unchecked(size, align);
-        let ptr = Global.alloc(layout);
-        if !ptr.is_null() {
-            ptr as *mut u8
-        } else {
-            oom()
-        }
+        Global.alloc(layout).unwrap_or_else(|_| oom()).as_ptr() as *mut u8
     }
 }
 
 #[cfg_attr(not(test), lang = "box_free")]
 #[inline]
 pub(crate) unsafe fn box_free<T: ?Sized>(ptr: Unique<T>) {
-    let ptr = ptr.as_ptr();
-    let size = size_of_val(&*ptr);
-    let align = min_align_of_val(&*ptr);
+    let size = size_of_val(ptr.as_ref());
+    let align = min_align_of_val(ptr.as_ref());
     // We do not allocate for Box<T> when T is ZST, so deallocation is also not necessary.
     if size != 0 {
         let layout = Layout::from_size_align_unchecked(size, align);
-        Global.dealloc(ptr as *mut Opaque, layout);
+        Global.dealloc(NonNull::new_unchecked(ptr.as_ptr() as *mut Opaque), layout);
```

Comment:
This can be `NonNull::from(ptr).cast()`

Reply:
> This can be NonNull::from(ptr).cast()

That's true in many other places. Same for things like `ptr.as_ptr() as *mut u8` which can probably be `ptr.cast().as_ptr()`

Reply:
My comment was mostly about using the safe `From` impl instead of the unsafe `new_unchecked`. I don’t have a strong opinion on `cast()` v.s. `as`.